README.md
# WEK WEK CODE - Extension

Wek2 Code Extension is Extension for Template Like Bootstrap Template, CodeIgniter, HTML Starter Template, and many more (will be update soon)

## Features

- Duck Starter Button
  - Bootstrap 5.3 Template With Dark And Lightmode System
  - CodeIgniter 4 Template
  - Start PHP Server with port localhost:8080

### Feature Key Binding

- Select Link CSS Or JS whene You In the text Editor and press CTRL+SHIFT+D , and files will be download automaticaly in your folder

## Command Feature

- CI4 Controller | To make New Controller
- CI4 Route | Make New Lines Code Of Routes.php file
- CI4 Template | Make New Project for Codeigniter4

More Feature Will Be Update Soon

## Installation

- Go to Extension Tab In VsCode
- Search 'Wek Wek Code'
- Click Install

## Authors

- [@teguhsaint](https://github.com/teguhsaint)
- [@noval_amirul](https://instagram.com/noval_amirul)


CHANGELOG.md
# Change Log

All notable changes to the "wekwekcode" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

lib/test_connect.js
const vscode = require("vscode");
const mysql = require("mysql");

function get_set() {
  const myset = vscode.workspace.getConfiguration("wekwekcode");

  // membuat koneksi ke database
  const connection = mysql.createConnection({
    host: myset.get("mysqlhost"),
    user: myset.get("mysqlusername"),
    password: myset.get("mysqlpassword"),
    database: "",
  });
  connection.query("SHOW DATABASES", (error, results, fields) => {
    if (error) {
      console.error(error);
    } else {
      const items = results.map((result) => result.Database);
      vscode.window.showQuickPick(items).then((selectedItem) => {
        if (selectedItem) {
          // menggunakan database yang dipilih
          connection.changeUser({ database: selectedItem }, (error) => {
            if (error) {
              console.error(error);
            } else {
              // menjalankan query SQL untuk menampilkan daftar tabel
              connection.query("SHOW TABLES", (error, results2, fields) => {
                if (error) {
                  console.error(error);
                } else {
                  console.log("Daftar tabel:");
                  const tables = results2.map((result2) => result2["Tables_in_" + selectedItem]);
                  vscode.window.showQuickPick(tables).then((selectedTables) => {
                    if (selectedTables) {
                      console.log(`Tabel yang dipilih: ${selectedTables}`);
                    } else {
                      console.log("Tidak ada tabel yang dipilih");
                    }
                  });
                }
                connection.end();
              });
            }
          });
        }
      });
    }
  });
}
module.exports = {
  get_set,
};


lib/ci4_route.js
const vscode = require("vscode");
const path = require("path");
const direktori_user = vscode.workspace.workspaceFolders[0].uri.path.substring(1, 1000);
const fs = require("fs");
const parser = require("php-parser");

function tambah_route() {
  const routes_path = path.join(direktori_user, "app/Config/Routes.php");

  async function tambah() {
    const route_path = await vscode.window.showInputBox({ title: "Masukkan link tujuan" });

    const dirPath = path.join(direktori_user, "app/Controllers/");

    fs.readdir(dirPath, async (err, files) => {
      if (err) {
        vscode.window.showErrorMessage("Failed to read directory.");
        return;
      }

      const fileNames = files.map((file) => {
        return path.parse(file).name;
      });

      await vscode.window.showQuickPick(fileNames).then(async (selectedFileName) => {
        if (selectedFileName) {
          let filePath = path.join(direktori_user, "app/Controllers", selectedFileName + ".php");

          const fileContent = fs.readFileSync(filePath, "utf8");
          let namaClass = "";
          const ast = parser.parseCode(fileContent);
          const myfunc = [];
          ast.children.forEach((node) => {
            node.children.forEach((node) => {
              namaClass = node.name.name;
              node.body.forEach((node) => {
                myfunc.push(node.name.name);
              });
            });
          });
          await vscode.window.showQuickPick(myfunc).then((selectFunction) => {
            const new_route = `$routes->get('/${route_path}','${namaClass}::${selectFunction}');`;
            if (!route_path) {
              vscode.window.showErrorMessage("Path route tidak boleh kosong");
            } else {
              fs.readFile(routes_path, "utf8", function (err, data) {
                if (err) {
                  console.log(err);
                  vscode.window.showErrorMessage("Gagal membaca file routes.php");
                } else {
                  const target_line = `$routes->get('/', 'Home::index');`;
                  const lines = data.split("\n");
                  const target_index = lines.findIndex((line) => line.includes(target_line));
                  if (target_index === -1) {
                    vscode.window.showErrorMessage("Tidak dapat menemukan baris target");
                    return;
                  }
                  lines.splice(target_index + 1, 0, new_route);
                  fs.writeFile(routes_path, lines.join("\n"), function (err) {
                    if (err) {
                      console.log(err);
                      vscode.window.showErrorMessage("Gagal menambahkan route");
                    } else {
                      vscode.window.showInformationMessage("Berhasil menambahkan route");
                    }
                  });
                }
              });
            }
          });
        }
      });
    });
  }

  tambah();
}

module.exports = {
  tambah_route,
};


lib/ci4_controller.js
const vscode = require("vscode");
const path = require("path");
const direktori_user = vscode.workspace.workspaceFolders[0].uri.path.substring(1, 1000);
const fs = require("fs");

function cicontroller_func() {
  const dir_controller = path.join(direktori_user, "app/Controllers/");

  async function hh() {
    const nama_file = await vscode.window.showInputBox({ placeHolder: "contoh : CtrPelanggan", title: "Nama File Controller tanpa .php tidak boleh ada spasi" });
    function capitalize(s) {
      return s[0].toUpperCase() + s.slice(1);
    }
    if (!nama_file) {
      vscode.window.showErrorMessage("Nama Tidak Boleh Kosong");
    } else {
      const file_Controller = fs.createWriteStream(dir_controller + "/" + capitalize(nama_file.replaceAll(" ", "")) + ".php", (err) => {
        if (err) {
          console.log(err);
        }
      });

      const isi_controller = `<?php

namespace App\\Controllers;

class ${capitalize(nama_file.replaceAll(" ", ""))} extends BaseController
{
    public function index()
    {
        // return view('');
    }
}
`;
      file_Controller.write(isi_controller, (err) => {
        if (err) {
          file_Controller.close();
          console.log(err);
        } else {
          file_Controller.close();
          vscode.window.showInformationMessage("Done " + nama_file);
          const openPath = vscode.Uri.file(path.join(dir_controller, capitalize(nama_file) + ".php"));
          vscode.commands.executeCommand("vscode.open", openPath);
        }
      });
    }
  }
  hh();
}

module.exports = {
  cicontroller_func,
};


lib/download_template.js
const vscode = require("vscode");
const path = require("path");
const https = require("https");
const fs = require("fs");
const decompress = require("decompress");
const direktori_user = vscode.workspace.workspaceFolders[0].uri.path.substring(1, 1000);

function download_template_func(urlkus, statusbar) {
  statusbar.hide();

  let st2 = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 200);
  st2.text = "$(sync~spin) Duck is Busy..";
  st2.color = "white";
  st2.show();

  var urlku = urlkus;
  const basename = path.basename(urlku);
  const filzip = fs.createWriteStream(path.join(direktori_user, path.basename(urlku)));
  https.get(urlku, (response) => {
    vscode.window.setStatusBarMessage("Tunggu Sebentar.. Download Zip File...", 8500);

    var cur = 0;
    var len = parseInt(response.headers["content-length"], 10);

    response.on("data", function (chunk) {
      cur += chunk.length;
      vscode.window.setStatusBarMessage("Downloading " + ((100.0 * cur) / len).toFixed(2) + "%  Size : " + (cur / 1048576).toFixed(2) + " mb", 5000);
    });
    response.pipe(filzip);
    filzip.on("finish", (err) => {
      if (err) {
        console.log(err);
        return;
      }
      filzip.close();
      console.log("fone");

      vscode.window.withProgress(
        {
          location: vscode.ProgressLocation.Window,
          title: "Unzipping files...",
          cancellable: false,
        },
        async () => {
          try {
            const files = await decompress(direktori_user + "/" + basename, direktori_user)
              .then(() => {
                fs.unlink(direktori_user + "/" + basename, () => {});
                vscode.window.setStatusBarMessage("Selesai..", 5000);
              })
              .catch((error) => {
                console.log(error);
              });

            st2.hide();
            statusbar.show();
          } catch (error) {
            vscode.window.showErrorMessage(`Error unzipping files: ${error.message}`);
          }
        }
      );
    });
  });
}

module.exports = {
  download_template_func,
};


extension.js
const vscode = require("vscode");
const path = require("path");
const https = require("https");
const fs = require("fs");

// MODULE
const ci4_view = require("./lib/ci4_view");
const start_phpserver = require("./lib/start_phpserver");
const download_template = require("./lib/download_template");
const ci4_controller = require("./lib/ci4_controller");
const ci4_route = require("./lib/ci4_route");
// END MODULE

const direktori_user = vscode.workspace.workspaceFolders[0].uri.path.substring(1, 1000);
/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {
  let showS = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 200);
  showS.text = "$(my-duck)  Duck Starter";
  showS.command = "wekwekcode.duck_start";
  showS.color = "yellow";

  let duck_starter = vscode.commands.registerCommand("wekwekcode.duck_start", async function () {
    // array of options
    const opt2 = [
      {
        label: "Template",
        kind: vscode.QuickPickItemKind.Separator, // this is new
      },
      { label: "Bootstrap 5.3 Template", description: "Include Light & Dark Mode", items: 1 },
      { label: "CodeIgniter 4 Template", description: "Include extend & section views", items: 2 },

      {
        label: "CodeIgniter 4",
        kind: vscode.QuickPickItemKind.Separator, // this is new
      },
      { label: "Add Controller", description: "New CI4 Controller", items: 3 },
      { label: "Add Views", description: "New CI4 Views", items: 4 },
      { label: "Add Route", description: "New CI4 Route", items: 5 },

      {
        kind: vscode.QuickPickItemKind.Separator, // this is new
      },
      {
        kind: vscode.QuickPickItemKind.Separator, // this is new
      },
      { label: "Start PHP Server", items: 6 },
    ];

    // showQuickPick
    vscode.window.showQuickPick(opt2).then((selection) => {
      if (!selection) {
        // handle no selection
        return;
      }

      // handle selected option
      switch (selection.items) {
        case 1:
          download_template.download_template_func("https://binateknologi.com/tool/ex/bs53.zip", showS);
          break;
        case 2:
          download_template.download_template_func("https://binateknologi.com/tool/ex/ci4.zip", showS);
          break;
        case 3:
          vscode.commands.executeCommand("wekwekcode.ci_controller");
          break;
        case 4:
          vscode.commands.executeCommand("wekwekcode.ci_view_extend");
          break;
        case 5:
          vscode.commands.executeCommand("wekwekcode.ci_routes");
          break;
        case 6:
          start_phpserver.start_php_server();
          break;

        default:
          break;
      }
    });
  });

  let download_filess = vscode.commands.registerCommand("wekwekcode.downloadcss_orjs", function () {
    // Dapatkan teks yang dipilih pada editor
    const editor = vscode.window.activeTextEditor;
    const selectedText = editor.document.getText(editor.selection);

    // Salin teks yang dipilih ke clipboard sistem
    vscode.env.clipboard
      .writeText(selectedText)
      .then(() => {
        const copiedText = selectedText;
        const pattern = /http[s]?:\/\/(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+/g;
        const urls = copiedText.match(pattern);
        if (urls) {
          console.log(urls[0]);

          const download_files = fs.createWriteStream(path.join(direktori_user, path.basename(urls[0])));

          https.get(urls[0], (response) => {
            response.pipe(download_files);
          });
          download_files.on("finish", () => {
            download_files.close();

            vscode.env.clipboard.writeText(path.basename(urls[0]));
            vscode.commands.executeCommand("editor.action.clipboardPasteAction");
            vscode.window.setStatusBarMessage(`${path.basename(urls[0])} Has Been Downloaded`);
          });
        } else {
          console.log("Tidak ada URL dalam teks yang disalin");
          vscode.window.showErrorMessage("Tidak ada URL dalam teks yang disalin");
        }
      })
      .catch((err) => {
        console.error(err);
      });
  });

  let ci_controller = vscode.commands.registerCommand("wekwekcode.ci_controller", function () {
    ci4_controller.cicontroller_func();
  });

  let ci_routes = vscode.commands.registerCommand("wekwekcode.ci_routes", function () {
    ci4_route.tambah_route();
  });

  let ci_view_extend = vscode.commands.registerCommand("wekwekcode.ci_view_extend", async function () {
    const nama_file = await vscode.window.showInputBox({ placeHolder: "Masukkan Nama File Template" });
    if (nama_file) {
      ci4_view.buat_views(nama_file);
    }
  });

  context.subscriptions.push(showS, duck_starter, ci_controller, ci_routes, download_filess, ci_view_extend);
  showS.show();
}

function deactivate() {}

module.exports = {
  activate,
  deactivate,
};
